nexus/reconfigurator/simulation/src/system.rs

```diff
@@ -0,0 +1,473 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! A simulated reconfigurator system.
+
+use std::{collections::BTreeMap, sync::Arc};
+
+use chrono::Utc;
+use indexmap::IndexMap;
+use nexus_reconfigurator_planning::{
+    example::ExampleSystem, system::SystemDescription,
+};
+use nexus_types::{
+    deployment::Blueprint,
+    internal_api::params::{DnsConfigParams, DnsConfigZone},
+    inventory::Collection,
+};
+use omicron_common::api::external::Generation;
+use omicron_uuid_kinds::CollectionUuid;
+use uuid::Uuid;
+
+use crate::errors::{DuplicateError, KeyError, NonEmptySystemError};
+
+/// A versioned, simulated reconfigurator system.
+#[derive(Clone, Debug)]
+pub struct SimSystem {
+    // Implementation note: an alternative way to store data would be for
+    // `Simulator` to carry a global store with it, and then each system only
+    // stores the presence of blueprints/collections/etc rather than the
+    // objects themselves. In other words, a simulator-wide object store. A few
+    // things become easier that way, such as being able to iterate over all
+    // known objects of a given type.
+    //
+    // However, there are a few issues with this approach in practice:
+    //
+    // 1. The blueprints and collections are not guaranteed to be unique per
+    //    UUID. Unlike (say) source control commit hashes, UUIDs are not
+    //    content-hashed, so the same UUID can be associated with different
+    //    blueprints/collections.
+    // 2. DNS configs are absolutely not unique per generation! Again, not
+    //    content-hashed.
+    // 3. The mutable system may wish to not add objects to the store until
+    //    it's committed. This means that the mutable system would probably
+    //    have to maintain a list of pending objects to add to the store. That
+    //    complicates some of the internals, if not the API.
+    // 4. We'll have to figure out how to manage the store (so SimSystemBuilder
```

Comment:
It's unclear to me why any of this code would need to be multithreaded, especially if it's fully deterministic. The only use I could really see for multi-threading is computationally expensive things where the computation is deterministic. For things like that we could call out to something like `rayon` if necessary.